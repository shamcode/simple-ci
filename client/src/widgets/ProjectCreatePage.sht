{% import ErrorsPanel from './ErrorsPanel.sht' %}
{% import ProjectCreateForm from './ProjectCreateForm' %}

<ErrorsPanel errors={{errors}}/>

<ProjectCreateForm
        dataSaving={{dataSaving}}
        onSubmit={{this.createProject.bind( this )}}
/>