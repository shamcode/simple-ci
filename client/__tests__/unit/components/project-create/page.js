import { createDI } from 'sham-ui';
import ProjectCreatePage from '../../../../src/components/project-create/page.sfc';
import * as directives from 'sham-ui-directives';
import hrefto from 'sham-ui-router/lib/href-to';
import renderer from 'sham-ui-test-helpers';

const flushPromises = () => new Promise( resolve => setImmediate( resolve ) );

afterEach( () => {
    jest.resetModules();
    jest.clearAllMocks();
} );

it( 'renders correctly', () => {
    const DI = createDI();
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValue( '/' )
    } );

    const meta = renderer( ProjectCreatePage, {
        DI,
        directives: {
            ...directives,
            hrefto
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'create', () => {
    const DI = createDI();
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValue( '/' ),
        navigate: jest.fn()
    } );
    const createProjectMock = jest.fn();
    DI.bind( 'store', {
        createProject: createProjectMock.mockReturnValueOnce( Promise.resolve() )
    } );

    const { component: { container } } = renderer( ProjectCreatePage, {
        DI,
        directives: {
            ...directives,
            hrefto
        }
    } );
    const formData = {
        name: 'Test name',
        cwd: 'test cwd'
    };
    container.querySelector( '[name="name"]' ).value = formData.name;
    container.querySelector( '[name="cwd"]' ).value = formData.cwd;
    container.querySelector( '[type="submit"]' ).click();

    expect( createProjectMock.mock.calls ).toHaveLength( 1 );
    expect( createProjectMock.mock.calls[ 0 ] ).toHaveLength( 1 );

    const data = createProjectMock.mock.calls[ 0 ][ 0 ];
    expect( Object.keys( data ) ).toEqual( [ 'name', 'cwd' ] );
    expect( data.name ).toBe( formData.name );
    expect( data.cwd ).toBe( formData.cwd );
} );

it( 'create fail', async() => {
    expect.assertions( 3 );

    const DI = createDI();
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValue( '/' ),
        navigate: jest.fn()
    } );
    const createProjectMock = jest.fn();
    DI.bind( 'store', {
        createProject: createProjectMock.mockReturnValueOnce( Promise.reject() )
    } );

    const meta = renderer( ProjectCreatePage, {
        DI,
        directives: {
            ...directives,
            hrefto
        }
    } );
    const formData = {
        name: 'Test name',
        cwd: 'test cwd'
    };
    const { component: { container } } = meta;
    container.querySelector( '[name="name"]' ).value = formData.name;
    container.querySelector( '[name="cwd"]' ).value = formData.cwd;
    container.querySelector( '[type="submit"]' ).click();

    await flushPromises();

    expect( createProjectMock.mock.calls ).toHaveLength( 1 );
    expect( createProjectMock.mock.calls[ 0 ] ).toHaveLength( 1 );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
