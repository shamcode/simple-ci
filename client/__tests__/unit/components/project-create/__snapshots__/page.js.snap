// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create fail 1`] = `
Object {
  "Constructor": "ProjectCreatePage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "createProject": [Function],
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "errors": Array [
      "Create project fail!",
    ],
    "types": Array [],
  },
  "html": "
<ul class=\\"breadcrumb\\">
  <li>
    <a href=\\"/\\"> Projects </a>
  </li>
  <!--if-->
  <li>Create project</li>
</ul>
<!--BreadCrumb-->
<!--breadcrumbs-->
<ul class=\\"errors-panel\\">
  <li>Create project fail!</li>
</ul>
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Create</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
<!--content-->
<!--if-->
<!--PageContent-->
",
}
`;

exports[`renders correctly 1`] = `
Object {
  "Constructor": "ProjectCreatePage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "createProject": [Function],
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "errors": Array [],
    "types": Array [],
  },
  "html": "
<ul class=\\"breadcrumb\\">
  <li>
    <a href=\\"/\\"> Projects </a>
  </li>
  <!--if-->
  <li>Create project</li>
</ul>
<!--BreadCrumb-->
<!--breadcrumbs-->
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Create</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
<!--content-->
<!--if-->
<!--PageContent-->
",
}
`;
